### ISSUE TITLE
DSA configuration is not properly initialized

### ISSUE THREAD

**Issue by Comphix** [spdk/spdk#2486]
# Bug report

When testing the feature of https://review.spdk.io/gerrit/c/spdk/spdk/+/9304, it is found there is some error happened when data is persisted to PMem through DSA.

Looking at DSA device's status:

```
# ./scripts/setup.sh status
Hugepages
node     hugesize     free /  total
node0   1048576kB        0 /      0
node0      2048kB      843 /   1024
node1   1048576kB        0 /      0
node1      2048kB        0 /      0
 
Type     BDF             Vendor Device NUMA    Driver           Device     Block devices
NVMe     0000:27:00.0    8086   0953   0       vfio-pci         -          -
IDXD     0000:6a:01.0    8086   0b25   0       vfio-pci         -          -
IDXD     0000:6f:01.0    8086   0b25   0       vfio-pci         -          -
IDXD     0000:74:01.0    8086   0b25   0       vfio-pci         -          -
IDXD     0000:79:01.0    8086   0b25   0       vfio-pci         -          -
IDXD     0000:e7:01.0    8086   0b25   1       vfio-pci         -          -
IDXD     0000:ec:01.0    8086   0b25   1       vfio-pci         -          -
IDXD     0000:f1:01.0    8086   0b25   1       vfio-pci         -          -
IDXD     0000:f6:01.0    8086   0b25   1       vfio-pci         -          -
 
# lspci -vvvvvxxxx -s 0000:f6:01.0 | more
f6:01.0 System peripheral: Intel Corporation Device 0b25
        Subsystem: Intel Corporation Device 0000
……
        Capabilities: [160 v1] Transaction Processing Hints
                Device specific mode supported
                Steering table in TPH capability structure
……
150: 18 00 01 16 00 00 00 00 00 00 00 00 00 00 00 00
160: 17 00 01 17 05 02 01 00 00 00 00 00 00 00 00 00
170: 02 00 01 20 11 00 00 00 01 00 00 00 00 00 00 00
```

It looks like below content is incorrect 
```
160: 17 00 01 17 05 02 01 00 00 00 00 00 00 00 00 00
```
## Expected Behavior

The expected data is

```
Offset 168 should be 0x00000102.
Offset 16c should be 0x000A0000.
```

## Current Behavior

The current content is:

```
160: 17 00 01 17 05 02 01 00 00 00 00 00 00 00 00 00
```

## Steps to Reproduce

1. Reboot the system, check the DSA setting
```
The TPH looks like correct 
160: 17 00 01 17 05 02 01 00 02 01 00 00 00 00 0a 00
```
2. ./scripts/setup.sh
3. ./test/bdev/bdevperf/bdevperf -q 1 -t 2 -w write -o 65536 -m 0xF -C -z --wait-for-rpc
4. Check the DSA setting again:
```
The TPH looks like correct 
160: 17 00 01 17 05 02 01 00 02 01 00 00 00 00 0a 00
```
 5. ./scripts/rpc.py framework_start_init
6. Check the DSA again:
```
The content became all zero
160: 17 00 01 17 05 02 01 00 00 00 00 00 00 00 00 00
```

**Comment by optimistyzy** [comment]
I think when this IDXD_FLAG_DEST_STEERING_TAG is used, it does not work.
https://review.spdk.io/gerrit/c/spdk/spdk/+/12351/2/include/spdk/idxd.h

**Comment by peluse** [comment]
> I think when this IDXD_FLAG_DEST_STEERING_TAG is used, it does not work. https://review.spdk.io/gerrit/c/spdk/spdk/+/12351/2/include/spdk/idxd.h

Thanks Ziye.  it's a little confusing as we reversed the polarity of the bit wrt how its defined in the spec.  It is indeed being set correctly but what Gang has found, adn I've yet to root cause as I haven't debugged it yet, is that a PCI register that enables a PCIE feature for steering tags is apparently being overwritten during init and diableing the ability to use steering tags.  Stay tuned!

**Comment by peluse** [comment]
The values in offset 160 are being zeroed by this ioctl issued by DPDK during init (in eal_vfio.c).  Unclear as to how that's happening looking at the kernel code for handling this ioctl.  It is clear though, sitting on line 962 lspci shows the values at boot time.  Stepping over this one LOC results in the changes reported by @Comphix 

│   961                 /* get a file descriptor for the device */                                                                                             │
│B+ 962                 *vfio_dev_fd = ioctl(vfio_group_fd, VFIO_GROUP_GET_DEVICE_FD, dev_addr);                                                               │
│  >963                 if (*vfio_dev_fd < 0) {                                                                                                                │
│   964                         /* if we cannot get a device fd, this implies a problem with                                                                   │
│   965                          * the VFIO group or the container not having IOMMU configured.                                                                │
│   966                          */

**Comment by peluse** [comment]
FYI IOCLT handler is here: https://github.com/torvalds/linux/blob/master/drivers/vfio/vfio.c

**Comment by peluse** [comment]
Potential workout is WIP...

**Comment by peluse** [comment]
@Comphix not looking good on a workaround w/user space driver as this looks to be a VFIO kernel driver issue.  Still chasing down a few things.  In the meantime though, if you use the kernel driver option for idxd things should work, I've verified that much at least.

**Comment by gaowayne** [comment]
@peluse thank you so much Paul to find this so early, I did not see my accel_perf and bdevperf report any error. but I follow your steps my 160 offset are zeroed out too. May I know the impact for this error? it will cause data corruption?  
```
160: 17 00 01 17 05 02 01 00 00 00 00 00 00 00 00 00
170: 02 00 01 20 11 00 00 00 01 00 00 00 00 00 00 00
```

**Comment by peluse** [comment]
> @peluse thank you so much Paul to find this so early, I did not see my accel_perf and bdevperf report any error. but I follow your steps my 160 offset are zeroed out too. May I know the impact for this error? it will cause data corruption?
> 
> ```
> 160: 17 00 01 17 05 02 01 00 00 00 00 00 00 00 00 00
> 170: 02 00 01 20 11 00 00 00 01 00 00 00 00 00 00 00
> ```

The impact on the use of the cache flag (nontemporal) is performance but it's not been characterized, basically all writes will be sent to cache regardless of whether you set it or not.  The steering tag, as I understand it, is simply "required" and I believe you will get an error if you don't set it.  That's what started this whole issue but I don't see mention of the error above as we didn't get the issue filled until it was understood that the error came from the missing TPH settings.  @Comphix can you provide more details on the error you are getting, I believe you and @gaowayne are testing the same thing but he didn't mention an error, just unexpected performance. It would not surprise me if not using the sterring tag flag could result in a mis-compare but I have no evidence of this.

**Comment by gaowayne** [comment]
@Comphix  @peluse here is my log with my device. I will debug more these two days. if bdevperf verify has no bug, it should be DSA issue. :(
```
[root@localhost drivers.storage.thirdparty.write-shaping-raid.wsr]# ./test/bdev/bdevperf/bdevperf --json ./perfwsrpmemdsa4G.json --json-ignore-init-errors -r /var/tmp/spdk.sock -m 7 -q 8 -o 4096 -w verify -t 60
[2021-07-20 14:49:45.775310] Starting SPDK v22.05-pre git sha1 4fa5b1ab0 / DPDK 21.11.0 initialization...
[2021-07-20 14:49:45.775454] [ DPDK EAL parameters: [2021-07-20 14:49:45.775473] bdevperf [2021-07-20 14:49:45.775486] --no-shconf [2021-07-20 14:49:45.775498] -c 7 [2021-07-20 14:49:45.775509] --huge-unlink [2021-07-20 14.775525] --log-level=lib.eal:6 [2021-07-20 14:49:45.775540] --log-level=lib.cryptodev:5 [2021-07-20 14:49:45.775556] --log-level=user1:6 [2021-07-20 14:49:45.775571] --base-virtaddr=0x200000000000 [2021-07-20 14:49:45.7755match-allocations [2021-07-20 14:49:45.775598] --file-prefix=spdk_pid13922 [2021-07-20 14:49:45.775614] ]
EAL: No available 1048576 kB hugepages reported
TELEMETRY: No legacy callbacks, legacy socket not created
[2021-07-20 14:49:45.870442] app.c: 603:spdk_app_start: *NOTICE*: Total cores available: 3
[2021-07-20 14:49:46.073867] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 1
[2021-07-20 14:49:46.074050] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 2
[2021-07-20 14:49:46.074211] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 0
[2021-07-20 14:49:46.079900] accel_engine_idxd_rpc.c:  70:rpc_idxd_scan_accel_engine: *NOTICE*: Enabling IDXD user-mode
[2021-07-20 14:49:46.107984] accel_engine.c: 548:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
[2021-07-20 14:49:47.473332] accel_engine_idxd.c: 422:accel_engine_idxd_init: *NOTICE*: Accel engine updated to use IDXD DSA engine.
[2021-07-20 14:50:03.265762] bdev_wsrpmem.c: 489:create_wsrpmem_disk: *NOTICE*:  wsrpmem spdk_mem_register at: 0x0x7ed7c0000000
Running I/O for 60 seconds...
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 0
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 1
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 2
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 3
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 4
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 5
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 6
   First dword expected 0x0 got 0x5a5a5a00
Buffer mismatch! Target: WSRPMEM0 Disk Offset: 7
   First dword expected 0x0 got 0x5a5a5a00
Job run time less than one microsecond, no performance data will be shown
 Job: WSRPMEM0 (Core Mask 0x1)
```

**Comment by gaowayne** [comment]
> @Comphix @peluse here is my log with my device. I will debug more these two days. if bdevperf verify has no bug, it should be DSA issue. :(
> 
> ```
> [root@localhost drivers.storage.thirdparty.write-shaping-raid.wsr]# ./test/bdev/bdevperf/bdevperf --json ./perfwsrpmemdsa4G.json --json-ignore-init-errors -r /var/tmp/spdk.sock -m 7 -q 8 -o 4096 -w verify -t 60
> [2021-07-20 14:49:45.775310] Starting SPDK v22.05-pre git sha1 4fa5b1ab0 / DPDK 21.11.0 initialization...
> [2021-07-20 14:49:45.775454] [ DPDK EAL parameters: [2021-07-20 14:49:45.775473] bdevperf [2021-07-20 14:49:45.775486] --no-shconf [2021-07-20 14:49:45.775498] -c 7 [2021-07-20 14:49:45.775509] --huge-unlink [2021-07-20 14.775525] --log-level=lib.eal:6 [2021-07-20 14:49:45.775540] --log-level=lib.cryptodev:5 [2021-07-20 14:49:45.775556] --log-level=user1:6 [2021-07-20 14:49:45.775571] --base-virtaddr=0x200000000000 [2021-07-20 14:49:45.7755match-allocations [2021-07-20 14:49:45.775598] --file-prefix=spdk_pid13922 [2021-07-20 14:49:45.775614] ]
> EAL: No available 1048576 kB hugepages reported
> TELEMETRY: No legacy callbacks, legacy socket not created
> [2021-07-20 14:49:45.870442] app.c: 603:spdk_app_start: *NOTICE*: Total cores available: 3
> [2021-07-20 14:49:46.073867] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 1
> [2021-07-20 14:49:46.074050] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 2
> [2021-07-20 14:49:46.074211] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 0
> [2021-07-20 14:49:46.079900] accel_engine_idxd_rpc.c:  70:rpc_idxd_scan_accel_engine: *NOTICE*: Enabling IDXD user-mode
> [2021-07-20 14:49:46.107984] accel_engine.c: 548:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
> [2021-07-20 14:49:47.473332] accel_engine_idxd.c: 422:accel_engine_idxd_init: *NOTICE*: Accel engine updated to use IDXD DSA engine.
> [2021-07-20 14:50:03.265762] bdev_wsrpmem.c: 489:create_wsrpmem_disk: *NOTICE*:  wsrpmem spdk_mem_register at: 0x0x7ed7c0000000
> Running I/O for 60 seconds...
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 0
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 1
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 2
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 3
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 4
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 5
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 6
>    First dword expected 0x0 got 0x5a5a5a00
> Buffer mismatch! Target: WSRPMEM0 Disk Offset: 7
>    First dword expected 0x0 got 0x5a5a5a00
> Job run time less than one microsecond, no performance data will be shown
>  Job: WSRPMEM0 (Core Mask 0x1)
> ```

I reviewed bdevperf code, and also confirm P5800 verify works well. I do implement read on DSA well. let me confirm this by debug more.

**Comment by gaowayne** [comment]
@peluse actually my bdevperf perf is good but accel_perf perf is bad. now I have data consistent issue.

**Comment by peluse** [comment]
@gaowayne if PMEM is involved in these transfers I'd say no conclusions can be drawn from your results until we get the VFIO issue cleared up or until you try running with DSA using the kernel idxd driver instead of VFIO

**Comment by peluse** [comment]
@gaowayne  FYI the docs are missing an important step, see https://review.spdk.io/gerrit/c/spdk/spdk/+/12702

**Comment by gaowayne** [comment]
@peluse with latest SPDK accel_perf, I follow above steps to configure DSA kernel mode, it works it showed me I am using kernel driver, but accel_perf crashed. please see below error log from gdb, idxd is not created well
```
Starting program: /mnt/sda/sda5/wayne/dsa/spdkbase/spdk/build/examples/accel_perf -q 16 --json-ignore-init-errors -o 65536 -t 60 -w copy -c /root/wayne/dsa_kernel.json
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
SPDK Configuration:
Core mask:      0x1

Accel Perf Configuration:
Workload Type:  copy
Transfer size:  65536 bytes
Queue depth:    16
Allocate depth: 16
# threads/core: 1
Run time:       60 seconds
Verify:         No

[2022-05-20 17:53:05.737541] Starting SPDK v22.05-pre git sha1 494eb6e58 / DPDK 21.11.0 initialization...
[2022-05-20 17:53:05.737674] [ DPDK EAL parameters: [2022-05-20 17:53:05.737693] (null) [2022-05-20 17:53:05.737706] --no-shconf [2022-05-20 17:53:05.737717] -c 0x1 [2022-05-20 17:53:05.737729] --huge-unlink [2022-05-20 17:53:05.737745] --log-level=lib.eal:6 [2022-05-20 17:53:05.737759] --log-level=lib.cryptodev:5 [2022-05-20 17:53:05.737774] --log-level=user1:6 [2022-05-20 17:53:05.737792] --base-virtaddr=0x200000000000 [2022-05-20 17:53:05.737807] --match-allocations [2022-05-20 17:53:05.737818] --file-prefix=spdk_pid54992 [2022-05-20 17:53:05.737833] ]
[New Thread 0x7ffff75c8640 (LWP 54996)]
EAL: No available 1048576 kB hugepages reported
[New Thread 0x7ffff6dc7640 (LWP 54997)]
TELEMETRY: No legacy callbacks, legacy socket not created
[2022-05-20 17:53:05.830032] app.c: 603:spdk_app_start: *NOTICE*: Total cores available: 1
[2022-05-20 17:53:06.035118] reactor.c: 947:reactor_run: *NOTICE*: Reactor started on core 0
[2022-05-20 17:53:06.039508] accel_engine_idxd_rpc.c:  68:rpc_idxd_scan_accel_engine: *NOTICE*: Enabling IDXD kernel-mode
[2022-05-20 17:53:06.043543] accel_engine.c: 548:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
[2022-05-20 17:53:06.047155] accel_engine_idxd.c: 422:accel_engine_idxd_init: *NOTICE*: Accel engine updated to use IDXD DSA engine.
Running for 60 seconds...

Thread 1 "reactor_0" received signal SIGSEGV, Segmentation fault.
0x0000000000421fd3 in spdk_idxd_get_channel (idxd=0x799910) at idxd.c:120
120		chan->portal = idxd->impl->portal_get_addr(idxd);
(gdb) bt
#0  0x0000000000421fd3 in spdk_idxd_get_channel (idxd=0x799910) at idxd.c:120
#1  0x0000000000421965 in idxd_select_device (chan=0x76c100) at accel_engine_idxd.c:112
#2  idxd_create_cb (io_device=<optimized out>, ctx_buf=0x76c100) at accel_engine_idxd.c:346
#3  0x000000000042fe47 in spdk_get_io_channel (io_device=0x5571a0 <idxd_accel_engine>) at thread.c:2178
#4  0x000000000042bc5c in accel_engine_create_cb (io_device=<optimized out>, ctx_buf=0x76be40) at accel_engine.c:504
#5  0x000000000042fe47 in spdk_get_io_channel (io_device=io_device@entry=0x5574e0 <spdk_accel_module_list>) at thread.c:2178
#6  0x000000000042cabc in spdk_accel_engine_get_io_channel () at accel_engine.c:532
#7  0x000000000041f7be in _init_thread (arg1=0x798300) at accel_perf.c:654
#8  0x000000000042d965 in msg_queue_run_batch (max_msgs=<optimized out>, thread=0x99c940) at thread.c:716
#9  thread_poll (thread=thread@entry=0x99c940, max_msgs=max_msgs@entry=0, now=now@entry=18633245542749) at thread.c:932
#10 0x000000000042e95e in spdk_thread_poll (thread=thread@entry=0x99c940, max_msgs=max_msgs@entry=0, now=18633245542749) at thread.c:992
#11 0x0000000000428200 in _reactor_run (reactor=0x76b240) at reactor.c:924
#12 reactor_run (arg=0x76b240) at reactor.c:962
#13 0x00000000004286b7 in spdk_reactors_start () at reactor.c:1064
#14 0x0000000000425895 in spdk_app_start (opts_user=opts_user@entry=0x7fffffffde40, start_fn=start_fn@entry=0x41eda0 <accel_perf_start>, arg1=arg1@entry=0x0) at app.c:645
#15 0x000000000041c345 in main (argc=12, argv=0x7fffffffe028) at accel_perf.c:807
(gdb) p chan
$1 = (struct spdk_idxd_io_channel *) 0x76c140
(gdb) p *chan
$2 = {idxd = 0x799910, portal = 0x0, portal_offset = 0, batch = 0x0, desc_base = 0x0, ops_pool = {tqh_first = 0x0, tqh_last = 0x76c168}, ops_outstanding = {tqh_first = 0x0, tqh_last = 0x76c178}, ops_base = 0x0, 
  batch_pool = {tqh_first = 0x0, tqh_last = 0x76c190}, batch_base = 0x0}
(gdb) p *idxd
$3 = {impl = 0x0, portal = 0x0, socket_id = 0, num_channels = 0, total_wq_size = 128, chan_per_device = 8, num_channels_lock = {__data = {__lock = 1, __count = 0, __owner = 54992, __nusers = 1, __kind = 0, 
      __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = "\001\000\000\000\000\000\000\000\320\326\000\000\001", '\000' <repeats 26 times>, __align = 1}}
(gdb) 
```

**Comment by peluse** [comment]
@gaowayne make sure you are on master, this problem was recently addressed with https://review.spdk.io/gerrit/c/spdk/spdk/+/12641 which has been merged

**Comment by gaowayne** [comment]
@peluse  I merge all master spdk idxd and accel related codes. now my accel_perf copy DRAM to DRAM works fine with kernel DSA driver. but if I run bdevperf to copy on pmem I see below error. it looks like like DSA complain the virtual address that pmem registered on pmem devdax, before SPDK DSA driver does not complain this.

this fault address reported is what I registered from pmem. 
fault_addr = 139662672789504, {

```
(gdb) p status
$27 = 0
(gdb) n
1143			STAILQ_REMOVE_HEAD(&chan->ops_outstanding, link);
(gdb) n
1144			rc++;
(gdb) n
1146			if (spdk_unlikely(IDXD_FAILURE(op->hw.status))) {
(gdb) p * op
$28 = {hw = {status = 131 '\203', {result = 0 '\000', dif_status = 0 '\000'}, rsvd = 0, bytes_completed = 0, fault_addr = 139662672789504, {delta_rec_size = 0, crc32c_val = 0, {dif_chk_ref_tag = 0, 
        dif_chk_app_tag_mask = 0, dif_chk_app_tag = 0}, dif_ins_comp = {rsvd = 0, ref_tag = 0, app_tag_mask = 0, app_tag = 0}, dif_upd_comp = {src_ref_tag = 0, src_app_tag_mask = 0, src_app_tag = 0, dest_ref_tag = 0, 
        dest_app_tag_mask = 0, dest_app_tag = 0}, op_specific = '\000' <repeats 15 times>}}, cb_arg = 0x18800a0, cb_fn = 0x4d8555 <idxd_done>, batch = 0xa749b0, desc = 0x2000156d1b80, crc_dst = 0x0, parent = 0x0, 
  count = 1, link = {stqe_next = 0x2000156d0f60}}
(gdb) n
1147				status = -EINVAL;
(gdb) list
1142	
1143			STAILQ_REMOVE_HEAD(&chan->ops_outstanding, link);
1144			rc++;
1145	
1146			if (spdk_unlikely(IDXD_FAILURE(op->hw.status))) {
1147				status = -EINVAL;
1148				_dump_sw_error_reg(chan);
1149			}
1150	
1151			switch (op->desc->opcode) {
(gdb) b _dump_sw_error_reg
Breakpoint 7 at 0x4dbd81: file idxd.c, line 1112.
(gdb) bt
#0  spdk_idxd_process_events (chan=0xa7cc10) at idxd.c:1147
#1  0x00000000004d8bed in idxd_poll (arg=0xa74970) at accel_engine_idxd.c:297
#2  0x000000000051503d in thread_execute_poller (thread=0xa808f0, poller=0xa74d40) at thread.c:819
#3  0x00000000005154e5 in thread_poll (thread=0xa808f0, max_msgs=0, now=111346383635005) at thread.c:941
#4  0x0000000000515612 in spdk_thread_poll (thread=0xa808f0, max_msgs=0, now=111346383635005) at thread.c:992
#5  0x00000000004e69b5 in _reactor_run (reactor=0xa42700) at reactor.c:924
#6  0x00000000004e6aa3 in reactor_run (arg=0xa42700) at reactor.c:962
#7  0x00000000004e6e6b in spdk_reactors_start () at reactor.c:1064
#8  0x00000000004e3917 in spdk_app_start (opts_user=0x7fffffffdd50, start_fn=0x40bd3d <bdevperf_run>, arg1=0x0) at app.c:645
#9  0x000000000040c7eb in main (argc=16, argv=0x7fffffffdf18) at bdevperf.c:2157
```
```
(gdb) s
spdk_accel_submit_copy (ch=0xa812d0, dst=0x2000156e7f00, src=0x7f05c0000000, nbytes=32768, flags=0, cb_fn=0x42de5f <wsrpmem_done>, cb_arg=0x20000724eab8) at accel_engine.c:180
180		struct accel_io_channel *accel_ch = spdk_io_channel_get_ctx(ch);
```

**Comment by gaowayne** [comment]
@peluse with accel_perf that suppor pmem copy it can reproduce the same issue with spdk master code based.
```
(gdb) p * op
$1 = {hw = {status = 131 '\203', {result = 0 '\000', dif_status = 0 '\000'}, rsvd = 0, bytes_completed = 0, fault_addr = 139662673801216, {delta_rec_size = 0, crc32c_val = 0, {dif_chk_ref_tag = 0, 
        dif_chk_app_tag_mask = 0, dif_chk_app_tag = 0}, dif_ins_comp = {rsvd = 0, ref_tag = 0, app_tag_mask = 0, app_tag = 0}, dif_upd_comp = {src_ref_tag = 0, src_app_tag_mask = 0, src_app_tag = 0, dest_ref_tag = 0, 
        dest_app_tag_mask = 0, dest_app_tag = 0}, op_specific = '\000' <repeats 15 times>}}, cb_arg = 0x99d190, cb_fn = 0x421610 <idxd_done>, batch = 0x799740, desc = 0x2000008fde80, crc_dst = 0x0, parent = 0x0, 
  count = 1, link = {stqe_next = 0x2000008fd260}}
(gdb) bt
#0  spdk_idxd_process_events (chan=0x7996d0) at idxd.c:1134
#1  0x00000000004217d1 in idxd_poll (arg=0x799690) at accel_engine_idxd.c:297
#2  0x000000000042d096 in thread_execute_poller (poller=0x799ad0, thread=0x7991d0) at thread.c:819
#3  thread_poll (thread=thread@entry=0x7991d0, max_msgs=max_msgs@entry=0, now=now@entry=130612144118917) at thread.c:941
#4  0x000000000042e03e in spdk_thread_poll (thread=thread@entry=0x7991d0, max_msgs=max_msgs@entry=0, now=130612144118917) at thread.c:992
#5  0x00000000004279b0 in _reactor_run (reactor=0x768f40) at reactor.c:924
#6  reactor_run (arg=0x768f40) at reactor.c:962
#7  0x0000000000427e67 in spdk_reactors_start () at reactor.c:1064
#8  0x0000000000425045 in spdk_app_start (opts_user=opts_user@entry=0x7fffffffde20, start_fn=start_fn@entry=0x41ecd0 <accel_perf_start>, arg1=arg1@entry=0x0) at app.c:645
#9  0x000000000041c355 in main (argc=12, argv=0x7fffffffe008) at accel_perf.c:809
```

**Comment by peluse** [comment]
@gaowayne I'm not following your last two updates.  As this issue is about the configuration of DSA and you seem to be running other experiments can I ask that you open a new issue (and leave this one to track the known VFIO problem) and open a new one with - and this is really important - your specific configuration and exact steps to reproduce.

**Comment by peluse** [comment]
@Comphix quick update: we have a kernel driver patch that works.  More next week...

**Comment by Comphix** [comment]
Thanks Paul. Will rebuild the kernel with the fix and verify it then.

**Comment by peluse** [comment]
expect to submit kernel patch to address this issue no later than next week (going through an internal review real quick).

**Comment by peluse** [comment]
Breaking news! Due to some DSA spec changes we no longer need TPH config space settings so once the following patch lands (has new flag settings as a workaround) writes to PMEM can happen w/o a special kernel patch or the workaround of using the idxd kernel driver https://review.spdk.io/gerrit/c/spdk/spdk/+/13572 @gaowayne @Comphix

**Comment by Comphix** [comment]
Thanks Paul. This is great!

**Comment by gaowayne** [comment]
> Breaking news! Due to some DSA spec changes we no longer need TPH config space settings so once the following patch lands (has new flag settings as a workaround) writes to PMEM can happen w/o a special kernel patch or the workaround of using the idxd kernel driver https://review.spdk.io/gerrit/c/spdk/spdk/+/13572 @gaowayne @Comphix

thank you so much Paul. this is great. :) I understand now.

**Comment by peluse** [comment]
Although a kernel patch was submitted to preserve TPH config space settings, it is no longer needed.  There is a new recommended way to assure persistence on writes to PMEM that doesn't require TPH and that code is already in place.  So, this issue is no longer relevant.  Closing.
